tests/serializers/test_functions.py

```diff
@@ -324,3 +326,54 @@ def serialize_custom_mapping(value, serializer, _info):
     assert s.to_python({'a': 1, 'b': 2}, mode='json', include={'a'}) == 'a=1'
     assert s.to_json({'a': 1, 'b': 2}) == b'"a=1 b=2"'
     assert s.to_json({'a': 1, 'b': 2}, exclude={'b'}) == b'"a=1"'
+
+
+def test_function_model():
+    @dataclass
+    class Model:
+        x: int
+
+        def ser_x(self, v: Any, _) -> str:
+            return f'{v:_}'
+
+    s = SchemaSerializer(
+        core_schema.model_schema(
+            Model,
+            core_schema.typed_dict_schema(
+                {
+                    'x': core_schema.typed_dict_field(
+                        core_schema.int_schema(serialization=core_schema.function_plain_ser_schema_model(Model.ser_x))
+                    )
+                }
+            ),
+        )
+    )
+    assert s.to_python(Model(x=1000)) == {'x': '1_000'}
+
+
+def test_function_wrap_model():
+    @dataclass
+    class Model:
+        x: int
+
+        def ser_x(self, v: Any, serializer: core_schema.SerializeWrapHandler, _) -> str:
+            x = serializer(v)
+            return f'{x:_}'
```

Comment:
Having implemented this what comes across as a bit weird is that these functions accept `self` _and_ the value. If you have access to `self, why do you need us to pass in the value? Besides, if you get it from the model instance it will already be typed and it has a good name. It's pretty common for validators at least to duplicate the field name and type hint into the validator signature. Which makes me think it would make more sense to just not pass in the value here as a parameter.

```python
# if you want to customize serialization for a type you can use Annotated
# or create a custom type with __pydantic_core_serializer__ implemented
MyInt = Annotated[int, PlainSerializer(lambda x, _: x * 2)]

class Model(BaseModel):
   # if you don't need the model
   # and want to customize serialization for just this field (not the type)
   # you can do that using Annotated in the field type hint

   # you get just the field's value
    x: Annotated[int, PlainSerializer(lambda x, _: x * 2)]
    # and can nest it inside a generic type
    y: Dict[str, Annotated[int, PlainSerializer(lambda x, _: x * 2)]]
    # wrap serializer applied to a field
    z: Annotated[List[int], WrapSerializer(lambda x, ser, _: ser(item, idx) for idx, item in enumerate(x))]

   # but if you want the whole model, either because
   # you want to customize serialization based on some other attribute of the model
   # (e.g. internationalization)
   # you can use @serialize applied to an instance method

    # you get the whole model but only serialize one field
    # this gets applied at the `typed-dict-field` level
    @serializer("x")
    def ser_x(self, info: core_schema.SerializationInfo) -> str:
        return f'{self.x:_}'

    # you get the whole model and serialize the entire thing
    # this gets applied at the `model` schema level
    @serializer
    def ser_model(self, info: core_schema.SerializationInfo) -> str:
        return json.dumps(self.model_dict())

    # you can also use a class or instance method
    # which would be the same as using Annotated on the field
    @serializer("x")
    @staticmethod
    def ser_x(x: int, info: core_schema.SerializationInfo) -> str:
        return f'{self.x:_}'
```

Obviously we're not going to implement all of this for _this_ PR. But I think what would make sense is to change this PR to support two signatures:
- `(__self, __info)`
- `(__value, __info)`
The two signatures are compatible, we'll have to "choose" one from `pydantic` depending on (1) if it was done via `Annotated`, a custom type or the `@serialize` decorator and (2) in the case of the `@serialize` decorator we'll introspect to detect if the first argument is `self` or `slf` and if so assume it's an unbound instance method, else assume it's a static or class method that accepts just the value.

Reply:
@dmontagu your thoughts on this would be appreciated

Reply:
I don't have a very strong opinion about what should be done here, but I will say that I don't _fully_ buy into the "If you have access to `self`, why do you need us to pass in the value? Besides, if you get it from the model instance it will already be typed and it has a good name." line of reasoning — in general, I would expect interacting with `self` to be significantly less common than just interacting with the value alone, and I think it's okay if there's multiple ways of getting at the value if you do choose to interact with self.

I think ease-of-use/simplicity for users is significantly more important than "normalizing" the API, and to that end I'm inclined to try to reduce the number of supported signatures, rather than have extra logic/patterns for people to learn. I think my gut reaction is that the more options we expose, the more challenging it will be for typical users to get comfortable with the full spectrum of functionality (especially considering that, in my experience, most typical users are writing at most a handful of custom validators in their whole application).

Reply:
> in general, I would expect interacting with `self` to be significantly less common than just interacting with the value alone

> I'm inclined to try to reduce the number of supported signatures, rather than have extra logic/patterns for people to learn.

Just to be clear, I think we're always going to have to support two signatures. It's either going to be `(__self, __info)` and `(__value, __info)` or `(__self, __value, __info)` and `(__value, __info)`. Or are you suggesting something else?

Reply:
Maybe I'm not understanding; I'm mostly just concerned with making sure that it's hard to find yourself in a confusing situation as a user. When you say:

> I think what would make sense is to change this PR to support two signatures:
>
>    * (__self, __info)
>    * (__value, __info)

I see the double-leading-underscore and think this means these variables will be treated as positional, and my concern is that whatever logic gets used to determine which one is used could result in surprises for the user and/or be something that is hard to discover without reading the docs. (Maybe that's unavoidable but the more "discoverable" we can make this sort of thing, the better, I think. For example, `self` _could_ live on `info` — that would potentially make it more discoverable, though has the obvious shortcoming of losing type-hints. So I don't _actually_ think that's a good alternative, just making the point.)

Ultimately I think it's okay as long as users get early and clear feedback about what they've done wrong; I think it would be more of a problem if they don't find out the signature is wrong until the function gets called to validate data. (As opposed to during class initialization.)

Reply:
Having thought about this more I'm definitely more in favor of dropping `value` as an input if you are getting `self`. As long as that doesn't make the implementation more complicated. If it does, I'm fine with keeping `value` as an input to the function call.

Reply:
It won't be much different, I'll change this to be `(self, info)`